src/hotspot/share/jfr/leakprofiler/checkpoint/objectSampleCheckpoint.cpp

```diff
@@ -202,7 +202,7 @@ static void prepare_for_resolution() {
 
 static bool stack_trace_precondition(const ObjectSample* sample) {
   assert(sample != NULL, "invariant");
-  return sample->has_stack_trace_id() && !sample->is_dead();
+  return sample->has_stack_trace_id() && !sample->is_dead() && !sample->stacktrace().valid();
```

Comment:
I am curious - why is `!sample->stacktrace().valid()` required here?

Reply:
Hi Jaroslav, so this is another optimization to reduce the size:

The previous writer.copy() usage was because there needed to be a mechanism to ensure that the newest sample candidates also had their stacktraces serialized in the current chunk. Object samples are emitted (as part of recording dump/stop) before on_rotation(). The old object events would be emitted, but their stacktraces would be serialized in the next on_rotation() call. These "new candidate" stacktraces are not really needed in every chunk but only those that emit object samples. A way to reduce the number of stacktraces written in every chunk is to let the ObjectSampleCheckpoint::write() first create the needed stacktrace blobs right there before serializing them. Then, it will only lay down the blobs for the needed candidates, and the automatic serialization of object sample stacktraces (via writer.copy() which installs the data into the checkpoint queue) can be avoided. With this change, the upcoming on_rotation() will have to navigate state correctly since some of these new candidates already have stacktrace blobs associated (if there was a previous object sample emit call), and the system would assert. Hence the need to check if a stacktrace blob has already been created via the .valid() call.

Reply:
I would like to get this fixed in JDK18, so I will withdraw this PR and recreate it for JDK18 instead. Stay tuned.